pkg/dns/config/sync.go

```diff
@@ -0,0 +1,200 @@
+		},
+	}
+
+	store, controller := cache.NewInformer(
+		listWatch,
+		&api.ConfigMap{},
+		time.Duration(0),
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    sync.onAdd,
+			DeleteFunc: sync.onDelete,
+			UpdateFunc: sync.onUpdate,
+		})
+
+	sync.store = store
+	sync.controller = controller
+
+	return sync
+}
+
+// kubeSync implements Sync for the Kubernetes API.
+type kubeSync struct {
+	ns   string
+	name string
+
+	client     clientset.Interface
+	store      cache.Store
+	controller *cache.Controller
+
+	channel chan *Config
+
+	latestVersion string
+}
+
+var _ Sync = (*kubeSync)(nil)
+
+func (sync *kubeSync) Once() (*Config, error) {
+	cm, err := sync.client.Core().ConfigMaps(sync.ns).Get(sync.name)
+
+	if err != nil {
+		glog.Errorf("Error getting ConfigMap %v:%v err: %v",
+			sync.ns, sync.name, err)
+		return nil, err
+	}
+
+	config, _, err := sync.processUpdate(cm)
+	return config, err
+}
+
+func (sync *kubeSync) Periodic() <-chan *Config {
+	go sync.controller.Run(wait.NeverStop)
+	return sync.channel
+}
+
+func (sync *kubeSync) toConfigMap(obj interface{}) *api.ConfigMap {
+	cm, ok := obj.(*api.ConfigMap)
+	if !ok {
+		glog.Fatalf("Expected ConfigMap, got %T", obj)
+	}
+	return cm
+}
+
+func (sync *kubeSync) onAdd(obj interface{}) {
+	cm := sync.toConfigMap(obj)
+
+	glog.V(2).Infof("ConfigMap %s:%s was created", sync.ns, sync.name)
+
+	config, updated, err := sync.processUpdate(cm)
+	if updated && err == nil {
+		sync.channel <- config
+	}
+}
+
+func (sync *kubeSync) onDelete(_ interface{}) {
+	glog.V(2).Infof("ConfigMap %s:%s was deleted, reverting to default configuration",
+		sync.ns, sync.name)
+
+	sync.latestVersion = ""
+	sync.channel <- NewDefaultConfig()
+}
+
+func (sync *kubeSync) onUpdate(_, obj interface{}) {
+	cm := sync.toConfigMap(obj)
+
+	glog.V(2).Infof("ConfigMap %s:%s was updated", sync.ns, sync.name)
+
+	config, changed, err := sync.processUpdate(cm)
+
+	if changed && err == nil {
+		sync.channel <- config
+	}
+}
+
+func (sync *kubeSync) processUpdate(cm *api.ConfigMap) (config *Config, changed bool, err error) {
+	glog.V(4).Infof("processUpdate ConfigMap %+v", *cm)
+
+	if cm.ObjectMeta.ResourceVersion != sync.latestVersion {
+		glog.V(3).Infof("Updating config to version %v (was %v)",
+			cm.ObjectMeta.ResourceVersion, sync.latestVersion)
+		changed = true
+		sync.latestVersion = cm.ObjectMeta.ResourceVersion
+	} else {
+		glog.V(4).Infof("Config was unchanged (version %v)", sync.latestVersion)
+		return
+	}
+
+	config = &Config{}
+
+	if err = sync.updateFederations(cm, config); err != nil {
+		glog.Errorf("Invalid configuration, ignoring update")
+		return
+	}
+
+	if err = config.Validate(); err != nil {
+		glog.Errorf("Invalid onfiguration: %v (value was %+v), ignoring update",
+			err, config)
+		config = nil
+		return
+	}
+
+	return
+}
+
+func (sync *kubeSync) updateFederations(cm *api.ConfigMap, config *Config) (err error) {
+	if flagValue, ok := cm.Data["federations"]; ok {
```

Comment:
I should have been clearer - the pattern we're recommending for most cases is to read a single configmap key as a JSON blob, and parse that into a native struct, rather than a bunch of keys.

This lets you build arbitrary config structure, with lists and maps and types and whatnot.  That JSON blob is the one that should hold typemeta.